File: app/src/main/java/com/elishaazaria/sayboard/ime/recognizers/WhisperLocal.java
package com.elishaazaria.sayboard.ime.recognizers;

import android.content.res.AssetManager;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

import com.elishaazaria.sayboard.data.LocalModel;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.util.concurrent.Executor;


class WhisperLib {
    public static native long initContextFromInputStream(InputStream inputStream);
    public static native long initContextFromAsset(AssetManager assetManager, String assetPath);
    public static native long initContext(String modelPath);
    public static native void freeContext(long contextPtr);
    public static native void fullTranscribe(long contextPtr, float[] audioData);
    public static native int getTextSegmentCount(long contextPtr);
    public static native String getTextSegment(long contextPtr, int index);

    private static final String LOG_TAG = "LibWhisper";

    private boolean isArmEabiV7a() {
        return Build.SUPPORTED_ABIS[0].equals("armeabi-v7a");
    }

    private String cpuInfo() {
        try {
            BufferedReader br = new BufferedReader(new FileReader("/proc/cpuinfo"));
            StringBuilder sb = new StringBuilder();
            String line = br.readLine();
            while (line != null) {
                sb.append(line);
                sb.append(System.lineSeparator());
                line = br.readLine();
            }
            return sb.toString();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return "";
    }

    public WhisperLib() {
        String info = cpuInfo();
        if (isArmEabiV7a() && info.contains("vfpv4")) {
            Log.i(LOG_TAG, "Loading libwhisper_vfpv4.so");
            System.loadLibrary("whisper_vfpv4");
        } else {
            Log.i(LOG_TAG, "Loading libwhisper.so");
            System.loadLibrary("whisper");
        }
    }
}

public class WhisperLocal implements RecognizerSource {
    private final MutableLiveData<RecognizerState> stateLD = new MutableLiveData<>(RecognizerState.NONE);
    private final LocalModel localModel;

    private WhisperLib whisperLib;
    private long whisperContext;
    private WhisperRecognizer recognizer;

    public WhisperLocal(LocalModel localModel) {
        this.localModel = localModel;
        this.whisperLib = new WhisperLib();
        this.whisperContext = 0;
    }

    @Override
    public void initialize(Executor executor, Observer<RecognizerSource> onLoaded) {
        stateLD.postValue(RecognizerState.LOADING);

        if (whisperContext != 0) {
            modelLoaded(whisperContext);
        }

        Handler handler = new Handler(Looper.getMainLooper());
        executor.execute(() -> {
            Long context = WhisperLib.initContext(localModel.path);
            handler.post(() -> {
                modelLoaded(context);
                onLoaded.onChanged(this);
            });
        });
    }

    private void modelLoaded(long context) {
        this.whisperContext = context;
        stateLD.postValue(RecognizerState.READY);
        recognizer = new WhisperRecognizer(context, 16000.0f);
    }

    private static class WhisperRecognizer implements Recognizer {
        private final float sampleRate;
        private final long whisperContext;

        public WhisperRecognizer(long context, float sampleRate) {
            this.sampleRate = sampleRate;
            this.whisperContext = context;
        }

        @Override
        public float getSampleRate() {
            return sampleRate;
        }

        @Override
        public void reset() {
        }

        @Override
        public boolean acceptWaveForm(short[] buffer, int nread) {
            return false;
        }

        @Override
        public String getResult() {
            return "";
        }

        @Override
        public String getPartialResult() {
            return "";
        }

        @Override
        public String getFinalResult() {
            return "";
        }
    }

    @Override
    public Recognizer getRecognizer() {
        return recognizer;
    }

    @Override
    public void close(boolean freeRAM) {
        if (freeRAM) {
            if (whisperContext != 0) WhisperLib.freeContext(whisperContext);
            whisperContext = 0;
            stateLD.postValue(RecognizerState.CLOSED);
        } else {
            stateLD.postValue(RecognizerState.IN_RAM);
        }
    }

    @Override
    public LiveData<RecognizerState> getStateLD() {
        return stateLD;
    }

    @Override
    public int getErrorMessage() {
        return 0;
    }

    @Override
    public String getName() {
        return localModel.locale.getDisplayName();
    }
}

File: app/src/main/java/com/volla/vollaboard/ime/recognizers/WhisperLocal.kt
package com.volla.vollaboard.ime.recognizers

import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.volla.vollaboard.data.LocalModel
import com.volla.vollaboard.ime.recognizers.WhisperLib
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import java.util.Locale
import java.util.concurrent.Executor

internal class WhisperLib {
    private val isArmEabiV7a: Boolean
        get() = Build.SUPPORTED_ABIS[0] == "armeabi-v7a"
    private val isArmEabiV8a: Boolean
        get() = Build.SUPPORTED_ABIS[0] == "arm64-v8a"

    private fun cpuInfo(): String {
        try {
            val br = BufferedReader(FileReader("/proc/cpuinfo"))
            val sb = StringBuilder()
            var line = br.readLine()
            while (line != null) {
                sb.append(line)
                sb.append(System.lineSeparator())
                line = br.readLine()
            }
            return sb.toString()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return ""
    }

    init {
        val info = cpuInfo()
        if (isArmEabiV7a && info.contains("vfpv4")) {
            Log.i(LOG_TAG, "Loading libwhisper_vfpv4.so")
            System.loadLibrary("whisper_vfpv4")
        } else if (isArmEabiV8a && info.contains("fphp")) {
            Log.i(LOG_TAG, "Loading libwhisper_v8fp16_va.so")
            System.loadLibrary("whisper_v8fp16_va")
        } else {
            Log.i(LOG_TAG, "Loading libwhisper.so")
            System.loadLibrary("whisper")
        }
    }

    companion object {
        external fun initContext(modelPath: String?): Long
        external fun freeContext(contextPtr: Long)
        external fun acceptAudio(contextPtr: Long, audioData: FloatArray?): Boolean
        external fun transcribe(contextPtr: Long)
        external fun getTextSegmentCount(contextPtr: Long): Int
        external fun getTextSegment(contextPtr: Long, index: Int): String
        external fun getSampleRate(): Int
        private const val LOG_TAG = "LibWhisper"
    }
}

class WhisperLocal(private val localModel: LocalModel) : RecognizerSource {
    private val stateMLD = MutableLiveData(RecognizerState.NONE)
    override val stateLD: LiveData<RecognizerState>
        get() = stateMLD
    private val whisperLib: WhisperLib = WhisperLib()
    private var whisperContext: Long
    private var myRecognizer: WhisperRecognizer? = null
    override val recognizer: Recognizer
        get() = myRecognizer!!

    init {
        whisperContext = 0
    }

    override fun initialize(executor: Executor, onLoaded: Observer<RecognizerSource?>) {
        stateMLD.postValue(RecognizerState.LOADING)
        if (whisperContext != 0L) {
            modelLoaded(whisperContext)
            return
        }
        val handler = Handler(Looper.getMainLooper())
        executor.execute {

            // TODO: Initialize Whisper with the target languages and
            // Additional features
            val context = WhisperLib.initContext(localModel.path)
            handler.post {
                modelLoaded(context)
                onLoaded.onChanged(this)
            }
        }
    }

    private fun modelLoaded(context: Long) {
        whisperContext = context
        stateMLD.postValue(RecognizerState.READY)
        myRecognizer = WhisperRecognizer(context, WhisperLib.getSampleRate().toFloat())
    }

    override val closed: Boolean
        get() = myRecognizer == null

    override val addSpaces: Boolean
        get() = !listOf("ja", "zh").contains(localModel.locale.language)

    private class WhisperRecognizer(
        private val whisperContext: Long,
        override val sampleRate: Float
    ) : Recognizer {
        private var result = ""
        override fun reset() {
            result = ""
        }

        override fun acceptWaveForm(buffer: ShortArray?, nread: Int): Boolean {
            val floatBuffer = FloatArray(nread)
            for (i in buffer!!.indices) {
                floatBuffer[i] = buffer[i].toFloat() / 32767.0f
            }
            return WhisperLib.acceptAudio(whisperContext, floatBuffer)
        }

        override fun getResult(): String {
            WhisperLib.transcribe(whisperContext)
            val textSegmentCount = WhisperLib.getTextSegmentCount(whisperContext)
            // TODO: Use StringBuilder here
            for (i in 0 until textSegmentCount) {
                result += WhisperLib.getTextSegment(whisperContext, i)
                result += ' '
            }
            return result
        }

        override fun getPartialResult(): String {
            return ""
            /*
            WhisperLib.transcribe(whisperContext);
            int textSegmentCount = WhisperLib.getTextSegmentCount(whisperContext);
            // TODO: Use StringBuilder here
            for (int i = 0; i < textSegmentCount; i++) {
                result += WhisperLib.getTextSegment(whisperContext, i);
                result += ' ';
            }
            return result;
             */
        }

        override fun getFinalResult(): String {
            return result
        }

        override val locale: Locale?
            get() = null
    }


    override fun close(freeRAM: Boolean) {
        if (freeRAM) {
            if (whisperContext != 0L) WhisperLib.freeContext(whisperContext)
            whisperContext = 0
            stateMLD.postValue(RecognizerState.CLOSED)
        } else {
            stateMLD.postValue(RecognizerState.IN_RAM)
        }
    }

    override val errorMessage: Int
        get() = 0
    override val name: String
        get() = localModel.locale.displayName
}